cohttp-eio/src/cohttp_eio.mli

```diff
@@ -90,3 +92,94 @@ module Server : sig
 
   val not_found_handler : handler
 end
+
+module Client : sig
+  type response = Http.Response.t * Eio.Buf_read.t
+
+  type host = string * int
+  (** Represents a server host domain name and port, e.g. www.example.org:8080,
+      www.reddit.com *)
+
+  type scheme = [ `HTTP | `HTTPS ]
+  (** Represents HTTP connection schemes. *)
+
+  type 'a conn = scheme -> host -> Eio.Flow.two_way as 'a
+  (** [a 'conn] is a lambda [(fun scheme host -> ... flow)] where [scheme]
+      represents a HTTP scheme and [host] represents a server host domain name
+      or address.
+
+      [flow] is the Eio flow value returned by the lambda. *)
+
+  type 'a body_disallowed_call =
+    ?version:Http.Version.t ->
+    ?headers:Http.Header.t ->
+    'a conn ->
+    Uri.t ->
+    response
+  (** [body_disallowed_call] denotes HTTP client calls where a request is not
+      allowed to have a request body.
+
+      It is a lambda [fun ?version ?headers conn_fn uri -> .. response)]. The
+      [uri] represents a valid and full http uri, e.g.
+      http://www.example.org/hello?q=123
+
+      @raise invalid_arg If [uri] doesn't containt full and valid HTTP uri *)
+
+  type 'a body_allowed_call =
+    ?version:Http.Version.t ->
+    ?headers:Http.Header.t ->
+    ?body:Body.t ->
+    'a conn ->
+    Uri.t ->
+    response
+  (** [body_allowed_call] denotes HTTP client calls where a request can
+      optionally have a request body.
+
+      It is a lambda [fun ?version ?headers ?body conn_fn uri -> .. response)].
+      The [uri] represents a valid and full http uri, e.g.
+      http://www.example.org/hello?q=123
+
+      @raise invalid_arg If [uri] doesn't containt full and valid HTTP uri *)
+
+  (** {1 Generic HTTP call} *)
+
+  val call :
+    ?meth:Http.Method.t ->
+    ?version:Http.Version.t ->
+    ?headers:Http.Header.t ->
+    ?body:Body.t ->
+    'a conn ->
+    Uri.t ->
+    response
+
+  (** {1 HTTP Calls with Body Disallowed} *)
+
+  val get : 'a body_disallowed_call
+  val head : 'a body_disallowed_call
+  val delete : 'a body_disallowed_call
+
+  (** {1 HTTP Calls with Body Allowed} *)
+
+  val post : 'a body_allowed_call
+  val put : 'a body_allowed_call
+  val patch : 'a body_allowed_call
+
+  (** {1 Response Body} *)
+
+  val read_fixed : response -> string option
+  (** [read_fixed (response,reader)] is [Some bytes], where [bytes] is of length
+      [n] if "Content-Length" header is a valid integer value [n] in [response].
+      [reader] is updated to reflect that [n] bytes was read. *)
+
+  val read_chunked : response -> (Body.chunk -> unit) -> Http.Header.t option
```

Comment:
Can we add a default handler and also have a read function invoking the correct read with a default handler (a dumb one filling a buffer then returning is contents should do, we just need to document that it could hang forever).

Reply:
To clarify my comment. I like the explicitness here but we start getting different incompatible APIs between the clients. It would be nice to have a `Body.read` for eio that just gives you the string, while keeping these two functions there or in a `Lowlevel` submodule.

Reply:
Chunked reading is quite different (behavior  and HTTP RFC spec) from fixed body reading; which is why they are a different/separate API. Just as an example, in chunked encoding we return a set of headers after chunk reading is done. However, this is not the case with `fixed` body read. 

Chunked read is also meant to be read in chunks so as to be memory efficient, therefore - except perhaps as a quick demo - I think using a buffer is not to be recommended. 

Finally, the cohttpe-eio client module keeps the read/write API consistent with the server where there is a separate read for fixed and chunked.

Reply:
I understand. As I said I lilke the lowlevel interface exposing the difference, and in the past we needed to add an expert mode to cohttp to actually be able to do something similar in a as backward compatible way as possible.

However, for most cases you get few bytes/kilobytes of body to read no matter the transfer encoding, and you just want to get it out as a string, so I think that we should have an inefficient read function, a sort of equivalent of `Cohttp.Body.read`, that hides this detail from you and justg gets the string for you, does not metter if it is not efficient (who wants efficiency can call read_chunked or read_fixed)

Reply:
I can certainly understand the allure of such a `read` function. I myself tried to shoehorn the two read functions into one. I mean why have two `read` functions when one will do. However, the resulting `read` API turned out to be incorrect, imprecise and confusing(to the user). 

For example, 
1. `read_chunked` returns an updated set of headers as specified in https://datatracker.ietf.org/doc/html/rfc7230#section-4.1.3. However, this is not the case with `read_fixed` which just returns a string blob. We could recommend in the API documentation to ignore such returned value when the response is fixed body. However why do the very thing that you recommend against. `read_fixed` is the recommended API without any such adulteration. 
2. Each chunk may have chunk extensions (https://datatracker.ietf.org/doc/html/rfc7230#section-4.1.1) which can contain metadata about how such a chunk maybe interpreted/consumed. How should we interpret such an extension? One option could be to silently ignore such extensions and pretend they don't exist. However, IIUC OCaml eschews such practices and generally does not recommend it. 

The current API of `read_fixed` and `read_chunked` are the result of above consideration.

Reply:
Then we should clearly document how to use it properly. Can you add to the documentation an example showing how to check which function to use and a simple example of use of both functions?

Reply:
Ack.

Reply:
Added demo/test for reading and writing chunk encoding.

writing chunk-encoding - https://github.com/bikallem/ocaml-cohttp/blob/eio-client/cohttp-eio/tests/test_client.ml#L39
reading chunk-encoding - https://github.com/bikallem/ocaml-cohttp/blob/eio-client/cohttp-eio/tests/test_client.ml#L90

Reply:
It seems that `read_fixed` could be considered as a special case of `read_chunked`, where there is only one chunk and there are no extra headers.

So, perhaps `read_chunked` should work on non-chunked bodies too, just calling the chunk handler once. Might need renaming (e.g. to just `read`).

Reply:
@talex5 Aha! Do you mean convert `read` data into `Body.chunk` ? Possibly could work. Let me see.

Reply:
> Do you mean convert read data into Body.chunk ? Possibly could work. Let me see.

Yes. I don't see why the comment you linked is a problem for that:

1. "read_chunked returns an updated set of headers" : just return the unmodified headers in the fixed case
2. "Each chunk may have chunk extensions" : just return an empty set of extension for fixed bodies